### PR TITLE
Allow any host when fallback policy fails

### PR DESCRIPTION
**Pull request by rgs1**
We use the default subset as our fallback policy. However, if there's
a problem with our metadata it's possible for the default subset
to become empty. In this case, we'd like any host to be selected.

This change adds a `panic_mode_any` boolean option to LbSubsetConfig
to enable this behavior.

Signed-off-by: Raul Gutierrez Segales <rgs@pinterest.com>

### CODE REVIEW

File: api/envoy/api/v2/cds.proto [hunk 1]
```diff
@@ -385,6 +385,11 @@ message Cluster {
     // going to an individual locality if said locality is disproportionally affected by the
     // subset predicate.
     bool scale_locality_weight = 5;
+
+    // If true, when a fallback policy is configured and it fails to find a host this will
```
Comment (snowp) [inline api/envoy/api/v2/cds.proto]: I'd reword this to make it clear that it's the subset LB that is failing to find a host according to the fallback policy, right now it sounds like the policy itself is failing which seems a bit misleading.

File: include/envoy/upstream/load_balancer_type.h [hunk 2]
```diff
@@ -58,6 +58,11 @@ class LoadBalancerSubsetInfo {
    * fraction of hosts removed from the original host set.
    */
   virtual bool scaleLocalityWeight() const PURE;
+
+  /*
+   * @return bool whether to pick any host if the fallback policy fails to find a host.
```
Comment (snowp) [inline include/envoy/upstream/load_balancer_type.h]: maybe something like `whether to attempt to select a host from the entire cluster if host selection from the fallback subset fails`

**Comment by rgs1** [comment]
@snowp thanks for the quick review. I thought about making fallback policies chainable, but the end result is the same though with a more complicated implementation.

Besides `[{ policy: DEFAULT_SUBSET, default_subset: { ... } }, { policy: ANY_ENDPOINT }]`, which is equivalent to what we get with this, there is no other possible chain at the moment thus I discarded the idea.

File: source/common/upstream/subset_lb.cc [hunk 3]
```diff
@@ -35,6 +35,14 @@ SubsetLoadBalancer::SubsetLoadBalancer(
       scale_locality_weight_(subsets.scaleLocalityWeight()) {
   ASSERT(subsets.isEnabled());
 
+  if (subsets.panicModeAny()) {
+    HostPredicate predicate = [](const Host&) -> bool { return true; };
+
+    panic_mode_subset_.reset(new LbSubsetEntry());
```
[the remark this answers is not in the document]
Comment (rgs1) [reply source/common/upstream/subset_lb.cc]: Done.

File: source/common/upstream/subset_lb.cc [hunk 4]
```diff
@@ -35,6 +35,14 @@ SubsetLoadBalancer::SubsetLoadBalancer(
...
+    panic_mode_subset_->priority_subset_.reset(
```
[the remark this answers is not in the document]
Comment (rgs1) [reply source/common/upstream/subset_lb.cc]: Done.

File: api/envoy/api/v2/cds.proto (hunk 1 above)
Comment (rgs1) [reply api/envoy/api/v2/cds.proto]: Done, lmk if it sounds better now.

File: include/envoy/upstream/load_balancer_type.h (hunk 2 above)
Comment (rgs1) [reply include/envoy/upstream/load_balancer_type.h]: Done.

**Comment by snowp** [comment]
The possibility that came to mind was being able to define multiple different default subsets, like 
```
- fallback_policy: DEFAULT_SUBSET
   default_subset: subset1
- fallback_polciy: DEFAULT_SUBSET
   default_subset: subset2
```

but you're right that it would be more complicated to implement. 

That said, I think it might be worth it to do some extra work here in order to make this functionality more extensible in the future and to avoid a potential deprecation cycle. 

@envoyproxy/maintainers Thoughts?

**Comment by rgs1** [comment]
@snowp I see. Currently we only alow one default subset definition, so we could also make  `default_subset` a repeatable (a list):

https://github.com/envoyproxy/data-plane-api/blob/master/envoy/api/v2/cds.proto#L347

which would be another way of achieving this (by allowing a default subset with an empty definition, that would mean match any).

The other option would be to make it so that `fallback_policy` takes a list of subset definitions, instead of policy names.

My problem with both of these routes is that right now the subset LB is not that performant (see #3929), so encouraging users to create more subsets is something that needs some thought (at least for our scale).

**Comment by rgs1** [comment]
Hmm, having second thoughts on conflating the pending perf work for the subset LB with making fallback policies more extensible.

I think we can ignore the perf concern now, and go with a more general approach if that would be more consistent with our other config primitives and if it'll leave us in a better place for future development, etc.

**Comment by htuch** [comment]
A couple of observations/questions:
1. We can't upgrade singleton to repeated fields, as we break JSON wire compatibility, so we would need to do the field deprecation dance if we go to the default  subset chaining design.
2. Do we have any use cases in mind that require anything other than "panic any"?

**Comment by snowp** [comment]
Not being able to change it to a repeated field is a good point, so we'd have to go through a deprecation cycle regardless. To me that makes it reasonable to just go for the panic flag since nobody's actually asking for multiple fallbacks. 

If someone requests that in the future, we could deprecate the singleton field + the panic flag at the same time and represent them both with a list of fallbacks.

Does that sound reasonable @rgs1 ?

File: api/envoy/api/v2/cds.proto [hunk 5]
```diff
@@ -385,6 +385,13 @@ message Cluster {
     // going to an individual locality if said locality is disproportionally affected by the
     // subset predicate.
     bool scale_locality_weight = 5;
+
+    // If true, when a fallback policy is configured and its corresponding subset fails to find
+    // a host this will cause any host to be selected instead.
+    //
+    // This is useful when using the default subset as the fallback policy, given the default
+    // subset might become empty which would cause requests to fail unless this option is enabled.
```
Comment (snowp) [inline api/envoy/api/v2/cds.proto]: Being very pedantic here, even if you hit panic mode here you won't be guaranteed that the request won't fail. Both the cluster being empty or all hosts being unhealthy (with lb panic mode being disabled) can cause the request to still fail. Maybe word word it along the lines of "will attempt to select a host from the entire cluster"?

**Comment by rgs1** [comment]
@htuch we don't have a use case besides "panic any".

@snowp yep, sounds reasonable (there'll be a bit more than that tho, e.g.: we'll need to be able to attach names to key selectors so they can be referenced from policies, etc, etc). 

Thanks!

File: api/envoy/api/v2/cds.proto [hunk 6]
```diff
@@ -385,6 +385,13 @@ message Cluster {
...
+    bool panic_mode_any = 6;
```
Comment (snowp) [inline api/envoy/api/v2/cds.proto]: Thoughts on naming this something more identifiable? Would be nice be able to clearly differentiate this from the panic mode used by the LB. maybe `subset_panic_mode`?

File: docs/root/configuration/cluster_manager/cluster_stats.rst [hunk 7]
```diff
@@ -247,6 +247,7 @@ decisions. Stats are rooted at *cluster.<name>.* and contain the following stati
   lb_subsets_removed, Counter, Number of subsets removed due to no hosts
   lb_subsets_selected, Counter, Number of times any subset was selected for load balancing
   lb_subsets_fallback, Counter, Number of times the fallback policy was invoked
+  lb_subsets_fallback_panic, Counter, Number of times the fallback policy failed so any host was selected
```
Comment (snowp) [inline docs/root/configuration/cluster_manager/cluster_stats.rst]: maybe "Number of times the subset panic mode triggered"?

File: api/envoy/api/v2/cds.proto (hunk 5 above)
Comment (rgs1) [reply api/envoy/api/v2/cds.proto]: Done.

File: api/envoy/api/v2/cds.proto (hunk 6 above)
Comment (rgs1) [reply api/envoy/api/v2/cds.proto]: Hmm, we are within the `LbSubsetConfig` message already, so not sure we need subset again. I actually like `panic_mode_any` because it a) sorta is like the other panic mode people know about and b) vaguely reference the likeliness with the ANY fallback policy... 

But, do pushback if you feel strongly about this :-)

File: docs/root/configuration/cluster_manager/cluster_stats.rst (hunk 7 above)
Comment (rgs1) [reply docs/root/configuration/cluster_manager/cluster_stats.rst]: Done.

File: api/envoy/api/v2/cds.proto (hunk 6 above)
Comment (snowp) [reply api/envoy/api/v2/cds.proto]: I don't feel strongly about it, it being nested is probably good for now. My concern was about confusing the two panic modes precisely because they are fairly similar, but we can just refer to this as subset panic mode elsewhere so I think it's fine

File: docs/root/intro/version_history.rst [hunk 8]
```diff
@@ -50,6 +50,7 @@ Version history
 * upstream: added :ref:`degraded health value<arch_overview_load_balancing_degraded>` which allows
   routing to certain hosts only when there are insufficient healthy hosts available.
 * upstream: added support for host weighting and :ref:`locality weighting <arch_overview_load_balancing_locality_weighted_lb>` in the :ref:`ring hash load balancer <arch_overview_load_balancing_types_ring_hash>`, and added a :ref:`maximum_ring_size<envoy_api_field_Cluster.RingHashLbConfig.maximum_ring_size>` config parameter to strictly bound the ring size.
+* upstream: added configuration option to select any host when the fallback policy fails
```
Comment (snowp) [inline docs/root/intro/version_history.rst]: super nit: add period

Comment (rgs1) [reply docs/root/intro/version_history.rst]: done
